middleware/websocket/websocket.go

```diff
@@ -0,0 +1,220 @@
+	}
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		return 0, err
+	}
+	defer conn.Close()
+
+	cmd := exec.Command(config.Command, config.Arguments...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		panic(err) // TODO
+	}
+
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		panic(err) // TODO
+	}
+
+	metavars, err := buildEnv(cmd.Path, r)
+	if err != nil {
+		panic(err) // TODO
+	}
+
+	cmd.Env = metavars
+
+	if err := cmd.Start(); err != nil {
+		panic(err)
+	}
+
+	reader(conn, stdout, stdin)
+
+	return 0, nil // we shouldn't get here.
+}
+
+// buildEnv creates the meta-variables for the child process according
+// to the CGI 1.1 specification: http://tools.ietf.org/html/rfc3875#section-4.1
+// cmdPath should be the path of the command being run.
+// The returned string slice can be set to the command's Env property.
+func buildEnv(cmdPath string, r *http.Request) (metavars []string, err error) {
+	remoteHost, remotePort, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		return
+	}
+
+	serverHost, serverPort, err := net.SplitHostPort(r.Host)
+	if err != nil {
+		return
+	}
+
+	metavars = []string{
+		`AUTH_TYPE=`,      // Not used
+		`CONTENT_LENGTH=`, // Not used
+		`CONTENT_TYPE=`,   // Not used
+		`GATEWAY_INTERFACE=` + GatewayInterface,
+		`PATH_INFO=`,       // TODO
+		`PATH_TRANSLATED=`, // TODO
+		`QUERY_STRING=` + r.URL.RawQuery,
+		`REMOTE_ADDR=` + remoteHost,
+		`REMOTE_HOST=` + remoteHost, // Host lookups are slow - don't do them
+		`REMOTE_IDENT=`,             // Not used
+		`REMOTE_PORT=` + remotePort,
+		`REMOTE_USER=`, // Not used,
+		`REQUEST_METHOD=` + r.Method,
+		`REQUEST_URI=` + r.RequestURI,
+		`SCRIPT_NAME=` + cmdPath, // path of the program being executed
+		`SERVER_NAME=` + serverHost,
+		`SERVER_PORT=` + serverPort,
+		`SERVER_PROTOCOL=` + r.Proto,
+		`SERVER_SOFTWARE=` + ServerSoftware,
+	}
+
+	// Add each HTTP header to the environment as well
+	for header, values := range r.Header {
+		value := strings.Join(values, ", ")
+		header = strings.ToUpper(header)
+		header = strings.Replace(header, "-", "_", -1)
+		value = strings.Replace(value, "\n", " ", -1)
+		metavars = append(metavars, "HTTP_"+header+"="+value)
+	}
+
+	return
+}
+
+// reader is the guts of this package. It takes the stdin and stdout pipes
+// of the cmd we created in ServeWS and pipes them between the client and server
+// over websockets.
+func reader(conn *websocket.Conn, stdout io.ReadCloser, stdin io.WriteCloser) {
+	// Setup our connection's websocket ping/pong handlers from our const values.
+	conn.SetReadLimit(maxMessageSize)
+	conn.SetReadDeadline(time.Now().Add(pongWait))
+	conn.SetPongHandler(func(string) error { conn.SetReadDeadline(time.Now().Add(pongWait)); return nil })
+	go ticker(conn)
+
+	for {
+		msgType, r, err := conn.NextReader()
+		if err != nil {
+			if msgType == -1 {
+				return // we are done, as we got a close method.
+			}
+			panic(err) // TODO do something else here.
```

Comment:
ha. Yeah I agree. I tried to make sure to added TODO next to each panic so I won't forget to change them, when we decided on what we wanted to do. I agree, panic isn't usually ideal. I wonder if it would make since to have a cleanup method that disconnect with a websocket abnormal error (10006) I believe.